cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -41,22 +41,41 @@ endif()
 
 set(APACHE_MIRROR "")
 
-macro(get_apache_mirror)
+macro(maybe_drop_backup_urls URLS)
+  if(CMAKE_VERSION VERSION_LESS 3.7)
+    # ExternalProject doesn't support backup URLs;
+    # Feature only available starting in 3.7
+    list(GET ${URLS} 0 ${URLS})
+  endif()
+endmacro()
+
+macro(get_apache_mirrors)
```

Comment:
We've reached the limit on readability, just write a python script in build-support and shell out to this.

Reply:
If we use Python here, we require Python to build Apache Arrow C++ with bundled Apache Thrift. I don't like it because users (not developers) may need to install Python for Apache Arrow C++.

Reply:
I agree with @kou. If you want Python here, you could write a script to generate the cmake list to append here. But IMO we're way overengineering something that just needs to be a hardcoded list of mirrors.